// appium-remote-debugger/build/lib/remote-debugger-real-device.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _logger = _interopRequireDefault(require("./logger"));

var _remoteDebugger = require("./remote-debugger");

var _rpcClientRealDevice = _interopRequireDefault(require("./rpc-client-real-device"));

class RemoteDebuggerRealDevice extends _remoteDebugger.RemoteDebugger {
  constructor(opts = {}) {
    super(opts);
    this.udid = opts.udid;
    this.skippedApps = ['lockdownd'];
  }

  async connect() {
    this.setup();
    this.rpcClient = new _rpcClientRealDevice.default({
      bundleId: this.bundleId,
      platformVersion: this.platformVersion,
      isSafari: this.isSafari,
      host: this.host,
      port: this.port,
      socketPath: this.socketPath,
      specialMessageHandlers: this.specialCbs,
      messageProxy: this.remoteDebugProxy,
      logAllCommunication: this.logAllCommunication,
      logAllCommunicationHexDump: this.logAllCommunicationHexDump,
      socketChunkSize: this.socketChunkSize,
      udid: this.udid
    });
    await this.rpcClient.connect();

    try {
      const appInfo = await this.setConnectionKey();

      _logger.default.debug('Connected to application');

      return appInfo;
    } catch (err) {
      await this.disconnect();
      return null;
    }
  }

}

exports.default = RemoteDebuggerRealDevice;require('source-map-support').install();
